src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs

```diff
@@ -122,29 +125,71 @@ public JsonConverter GetConverter(Type typeToConvert)
             }
 
             DefaultJsonTypeInfoResolver.RootDefaultInstance();
-            return GetConverterInternal(typeToConvert);
+            return GetConverterFromTypeInfo(typeToConvert);
+        }
+
+        /// <summary>
+        /// Same as GetConverter but does not root converters
+        /// </summary>
+        internal JsonConverter GetConverterFromTypeInfo(Type typeToConvert)
+        {
+            if (_cachingContext == null)
+            {
+                if (_isLockedInstance)
+                {
+                    InitializeCachingContext();
+                }
+                else
+                {
+                    // We do not want to lock options instance here but we need to return correct answer
+                    // which means we need to go through TypeInfoResolver but without caching because that's the
+                    // only place which will have correct converter for JsonSerializerContext and reflection
+                    // based resolver. It will also work correctly for combined resolvers.
+                    JsonConverter? converter = GetJsonTypeInfoFromResolver(typeToConvert)?.Converter;
+                    converter ??= GetConverterFromOptionsOrReflectionConverter(typeToConvert);
+                    return converter;
+                }
+            }
+
+            return _cachingContext.GetOrAddConverter(typeToConvert);
         }
 
-        internal JsonConverter GetConverterInternal(Type typeToConvert)
+        private JsonConverter GetConverterFromTypeInfoOrOptionsNotCached(Type typeToConvert)
         {
-            // Only cache the value once (de)serialization has occurred since new converters can be added that may change the result.
-            if (_cachingContext != null)
+            JsonConverter? converter = GetOrAddJsonTypeInfoEx(typeToConvert, configured: false, doNotThrow: true)?.Converter;
+
+            // we can get here if resolver returned null but converter was added for the type
+            converter ??= GetConverterFromOptions(typeToConvert);
+
+            if (converter == null)
+            {
+                ThrowHelper.ThrowNotSupportedException_BuiltInConvertersNotRooted(typeToConvert);
+                return null!;
+            }
+
+            return converter;
+        }
+
+        private JsonConverter? GetConverterFromOptions(Type typeToConvert)
```

Comment:
In the current behaviour we want to return the built-in converters (or throw an exception if they haven't been rooted yet). This seems like a breaking change?

Reply:
it's not breaking change. If you call GetConverter directly it will root converter first and then call this method will behave as it was.

On the other hand if this is called when context is bound but converters are not rooted (it may be called implicitly) you will get an exception same way as you used to.